### PR TITLE
Create specific exception for when snapshots are in progress

### PR DESCRIPTION
**Pull request by talevy**
delete and close index actions threw IllegalArgumentExceptions when
attempting to run against an index that has a snapshot in progress.

This change introduces a dedicated SnapshotInProgressException for
these scenarios. This is done to explicitly signal to clients that
this is the reason the action failed, and it is a retryable error.

relates to #37541.

### CODE REVIEW

**Comment by talevy** [comment]
run gradle build tests 2

**Review by dakrone — APPROVED** [review]
This LGTM but should have @original-brownbear take a look before merging.

**Comment by talevy** [comment]
run gradle build tests 1

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotInProgressException.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotInProgressException.java]: It seems this constructor isn't used anywhere?

**Review by original-brownbear — APPROVED** [review]
LGTM (assuming CI goes green eventually), one question/NIT :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotInProgressException.java
Comment (talevy) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotInProgressException.java]: woops

**Comment by talevy** [comment]
thanks for the reviews @dakrone and @original-brownbear!

**Comment by talevy** [comment]
run gradle build tests 2

**Comment by DaveCTurner** [comment]
@talevy we intend to delete `ClusterAlreadyBootstrappedException` in #37463, and in merging master into that one I noticed a conflict with this PR which adds `SnapshotInProgressException`. I think since neither exists in a released version we should renumber `SnapshotInProgressException` as if `ClusterAlreadyBootstrappedException` never existed, and that's what I've done in #37463 to resolve the merge conflict, but I note that this PR will be backported so maybe it's going to keep things simpler if we keep the numbering as-is. WDYT?

**Comment by talevy** [comment]
@DaveCTurner I see that change. the plan sounds reasonable to me. I will wait for this change to go in before backporting

**Comment by DaveCTurner** [comment]
@talevy looks like this was backported now, but does the version need adjusting in the deserialiser?

https://github.com/elastic/elasticsearch/blob/170d7413d099a624d6d7ba7448ecd1b228d0f64c/server/src/main/java/org/elasticsearch/ElasticsearchException.java#L1013

I'm also wondering whether we're missing a test somewhere, as I would have expected the mixed-cluster tests to pick this up.

**Comment by talevy** [comment]
@DaveCTurner thanks for reminding me, I made a note to follow-up on this, but forgot. I noticed 
that this is not the only Exception that has these version discrepancy. TooManyBucketsException also is set to 7.0. I will open a PR. No clue about tests. we can discuss on the follow-up PR

**Pull request merged**
